ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -474,36 +482,98 @@ public Void visitFunctionCall(
 
       final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName());
 
-      // this context gets updated as we process non lambda arguments
-      final TypeContext currentTypeContext = expressionTypeContext.getCopy();
+      final List<SqlArgument> argTypesForFirstPass = new ArrayList<>();
 
-      final List<SqlArgument> argTypes = new ArrayList<>();
-
-      final boolean hasLambda = node.hasLambdaFunctionCallArguments();
-      for (final Expression expression : node.getArguments()) {
-        final TypeContext childContext = TypeContextUtil.contextForExpression(
-            expression, expressionTypeContext, currentTypeContext
-        );
-        process(expression, childContext);
-        final SqlType resolvedArgType = childContext.getSqlType();
+      final List<Expression> arguments = node.getArguments();
 
+      for (final Expression expression : arguments) {
```

Comment:
this loop can probably be it's own method

Reply:
In addition, though I understand we need to do similar two-pass loops on all three classes today, but could we consider at least extracting them into a single util function call?